### PR TITLE
Some cleanup to Dockerfile

### PR DESCRIPTION
**Pull request by xis19**
With this docker file, it is confirmed that building FDB does not need a network connection by adding

```
--network none
```

when running the container and build the code.

### CODE REVIEW

File: docker/centos7/Dockerfile [hunk 1]
```diff
@@ -444,25 +498,30 @@ RUN source /opt/rh/devtoolset-${DEVTOOLSET_VERSION}/enable && \
...
-    mkdir -p /opt/msgpack_3.3.0 && \
-    tar --strip-components 1 --no-same-owner --no-same-permissions --directory /opt/msgpack_3.3.0 -xf msgpack-3.3.0.tar.gz && \
-    cd /opt/msgpack_3.3.0 && \
-    cmake . && \
-    make && \
-    cd .. && \
-    rm -rf /tmp/*
+    mkdir -p ${SCRATCH_DIR}/msgpack-3.3.0 && \
+    tar --strip-components 1 --no-same-owner --no-same-permissions --directory ${SCRATCH_DIR}/msgpack-3.3.0 -xf msgpack-3.3.0.tar.gz && \
+    cd msgpack-3.3.0 && \
```
Comment (jzhou77) [inline docker/centos7/Dockerfile:513]: `cd ${SCRATCH_DIR}/msgpack-3.3.0`?

**Review by jzhou77 — APPROVED** [review]

File: docker/centos7/Dockerfile (hunk 1 above)
Comment (xis19) [reply docker/centos7/Dockerfile:513]: Sure

File: docker/centos7/Dockerfile [hunk 2]
```diff
@@ -2,9 +2,14 @@ FROM centos:7 as build
 
 ARG DEVTOOLSET_VERSION=11
 
-WORKDIR /tmp
+ENV SCRATCH_DIR=/tmp/scratch
```
Comment (sfc-gh-kmakino) [inline docker/centos7/Dockerfile:5]: For what ever reason this environment variable gets set empty, `rm -rf ${SCRATCH_DIR}/*` will cause an interesting result. What's the rationale behind this change? Was `/tmp` not good enough?

Comment (sfc-gh-kmakino) [reply docker/centos7/Dockerfile:5]: This is just a question. I like this PR. Thank you!

Comment (xis19) [reply docker/centos7/Dockerfile:5]: I do have some considerations here:

1. Ideally, when I build anything I should use `mkstemp` in order to avoid any pollution, i.e. isolate as much as possible. However, doing that is a lot more work.
2. I noticed that `/tmp` has some subdirectories that are not belonging to the steps, even though it seems deleting them will not cause any problem, not sure we should do this.
3. Using a custom scratch dir is a compromise to `mkstemp`. But it does isolate the build environment.
4. even `$SCRATCH_DIR` is '', the worst case is that the docker container, which is already an isolated environment, is not usable anymore, I feel this is not too much disastrous, as there is really nothing important/unrecoverable inside.

Please let me know if you have more concerns.

**Comment by xis19** [comment]
> I prefer to drop most of the changes in this PR. It does not appear to make any required improvements beyond the fix to the `addr2line` function.

I think I have reasons for most of the changes:

1. Moving LLVM(2000s) and FDB build(600s) to some earlier stages saves time when modifying later stages, which take much shorter time.
2. When thinking parallelism, using `$(nproc)` should be better than giving a specific CPU number while POSIX standard prefers using `$()` syntax over `\`\`` syntax.
3. Printing out the current operation using `echo` helps faster locate the layer when something goes wrong. Docker will print the command while in our scenario the first part of the command will mostly like `source dev tools; source python ...` and it would be really hard to know what it is doing immediately.
4. Moving comments to the lines they are referring should provide better readability than leaving them to the beginning of the `RUN` command.
5. Do out-of-source build instead of in-source build is the de factor style CMake does. It also reduces the size and the complexity of the generated images.

I agree this change has no functionality change, but I believe this cleanup would be more beneficial if merged. Please let me know your thoughts about this.
